Sources/Caching/KeyedDeferredValueStore.swift

```diff
@@ -0,0 +1,74 @@
+//
+//  Copyright RevenueCat Inc. All Rights Reserved.
+//
+//  Licensed under the MIT License (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      https://opensource.org/licenses/MIT
+//
+//  KeyedDeferredValueStore.swift
+//
+//  Created by Jacob Zivan Rakidzich on 8/12/25.
+
+import Foundation
+
+/// Some Task in which the value is Sendable and it can result in an error
+@_spi(Internal) public typealias AnyTask<T: Sendable> = Task<T, Error>
+
+/// Holds onto ``AnyTask`` objects by key, autoclearing on failure
+@_spi(Internal) public actor KeyedDeferredValueStore<H: Hashable, T: Sendable> {
+    var deferred: [H: AnyTask<T>] = [:]
+
+    /// Sets the task in the cache if one is not found
+    /// - Parameters:
+    ///   - task: The function that should execute if one is not found
+    ///   - key: The key to look up the result by
+    /// - Returns: The stored task
+    @_spi(Internal) public func getOrPut(
+        _ task: @escaping @Sendable @autoclosure () -> AnyTask<T>,
+        forKey key: H
+    ) -> AnyTask<T> {
+        guard let result = self.deferred[key] else {
+            let wrapped: AnyTask<T> = self.forgettingFailure(task, forKey: key)
+            self.deferred[key] = wrapped
```

Comment:
For long-running apps or scenarios with many unique tasks, it might be worth considering adding a cap on the size of the `deferred` dictionary to avoid this store from taking up too much memory. 

It looks like we remove entries for tasks that result in a failure, but since we never remove successful items, the growth potential of the deferred dictionary is uncapped. Once the size cap is reached, we could evict a previous entry, perhaps the least recently used entry?